sources/hybrid-prog-model.c

```diff
@@ -0,0 +1,164 @@
+  }
+}
+//<EG END ID="omp_thread">
+
+//<EG BEGIN ID="mpi_thread">
+static void parallel_region_MPI_cb(size_t evhdlr_registration_id,
+                                   pmix_status_t status,
+                                   const pmix_proc_t *source,
+                                   pmix_info_t info[], size_t ninfo,
+                                   pmix_info_t results[], size_t nresults,
+                                   pmix_event_notification_cbfunc_fn_t cbfunc,
+                                   void *cbdata) {
+  printf("Entered %s\n", __FUNCTION__);
+  /* do what we need MPI to do on entering a parallel region */
+  if (NULL != cbfunc) {
+    /* do what we need MPI to do on entering a parallel region */
+    cbfunc(PMIX_EVENT_ACTION_COMPLETE, NULL, 0, NULL, NULL, cbdata);
+  }
+}
+//<EG END ID="mpi_thread">
+
+static int openmp_handler() {
+  pmix_info_t *info;
+  int rc;
+
+  printf("Entered %s\n", __FUNCTION__);
+
+  //<EG BEGIN ID="omp_thread">
+  bool is_true = true;
+  pmix_status_t code = PMIX_OPENMP_PARALLEL_ENTERED;
+  PMIX_INFO_CREATE(info, 2);
+  PMIX_INFO_LOAD(&info[0], PMIX_EVENT_HDLR_NAME, "OpenMP-Master", PMIX_STRING);
+  PMIX_INFO_LOAD(&info[1], PMIX_EVENT_HDLR_FIRST, &is_true, PMIX_BOOL);
+  rc = PMIx_Register_event_handler(&code, 1, info, 2, parallel_region_OMP_cb, NULL, NULL);
+  if (rc < 0)
+    fprintf(stderr, "%s: Failed to register event handler for OpenMP region entrance\n", __FUNCTION__);
+  PMIX_INFO_FREE(info, 2);
+  //<EG END ID="omp_thread">
+  printf("Registered OpenMP event handler for OpenMP parallel region entered\n");
+
+  return rc;
+}
+
+static int mpi_handler() {
+  pmix_info_t *info;
+  int rc;
+
+  printf("Entered %s\n", __FUNCTION__);
+
+  //<EG BEGIN ID="mpi_thread">
+  pmix_status_t code = PMIX_OPENMP_PARALLEL_ENTERED;
+  PMIX_INFO_CREATE(info, 2);
+  PMIX_INFO_LOAD(&info[0], PMIX_EVENT_HDLR_NAME, "MPI-Thread", PMIX_STRING);
+  PMIX_INFO_LOAD(&info[1], PMIX_EVENT_HDLR_AFTER, "OpenMP-Master", PMIX_STRING);
+  rc = PMIx_Register_event_handler(&code, 1, info, 2, parallel_region_MPI_cb, NULL, NULL);
+  if (rc < 0)
+    fprintf(stderr, "%s: Failed to register event handler for OpenMP region entrance\n", __FUNCTION__);
+  PMIX_INFO_FREE(info, 2);
+  //<EG END ID="mpi_thread">
+  printf("Registered MPI event handler for OpenMP parallel region entered\n");
+
+  return rc;
+}
+
+static void notify_complete(pmix_status_t status, void *cbdata)
+{
+  volatile bool *flag = (volatile bool*)cbdata;
+  *flag = true;
+}
+
+int main() {
+  //<EG BEGIN ID="declare_model">
+  pmix_proc_t myproc;
+  pmix_info_t *info;
+  volatile bool wearedone = false;
+
+  PMIX_INFO_CREATE(info, 4);
+  PMIX_INFO_LOAD(&info[0], PMIX_PROGRAMMING_MODEL, "MPI", PMIX_STRING);
+  PMIX_INFO_LOAD(&info[1], PMIX_MODEL_LIBRARY_NAME, "FooMPI", PMIX_STRING);
+  PMIX_INFO_LOAD(&info[2], PMIX_MODEL_LIBRARY_VERSION, "1.0.0", PMIX_STRING);
+  PMIX_INFO_LOAD(&info[3], PMIX_THREADING_MODEL, "pthread", PMIX_STRING);
```

Comment:
I would not call "Pthreads" a threading model - user vs. system level is a threading model, but Pthreads merely an implementation, IMHO

Reply:
We should probably open a separate issue on this.  Currently, the example provided in the main part of the standard for this attribute is `pthreads`: https://github.com/pmix/pmix-standard/blob/ed7f00ccf1f3d0b76f42e782e3eff340fe3077c6/Chap_API_Init.tex#L253

That is where I got that key/value from.  Ultimately, I will defer to you/others on this since it is not my area of expertise.

Side-note: I do need to at least make it plural in my example to make the example earlier in the standard.